pkg/util/limiter/memory_limiter.go

```diff
@@ -0,0 +1,242 @@
+package limiter
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"go.uber.org/atomic"
+)
+
+var (
+	errCheckIntervalOutOfRange = errors.New(
+		"checkInterval must be greater than zero")
+	errLimitOutOfRange = errors.New(
+		"MemoryLimitBytes or ResetBytes must be greater than zero")
+)
+
+type MemoryLimiterConfig struct {
+	Enabled          bool          `yaml:"enabled"`
+	MemoryLimitBytes uint64        `yaml:"memory_limit_bytes"`
+	ResetBytes       uint64        `yaml:"reset_bytes"`
+	CheckInterval    time.Duration `yaml:"check_interval"`
+}
+
+// RegisterFlags registers flags.
+func (cfg *MemoryLimiterConfig) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	f.BoolVar(&cfg.Enabled, prefix+"memory-limiter.enabled", false, "Enable memory-limiter.")
+	f.DurationVar(&cfg.CheckInterval, prefix+"memory-limiter.check-interval", time.Second, "Configures how often to watch the heap to detect memory pressure.")
+	f.Uint64Var(&cfg.MemoryLimitBytes, prefix+"memory-limiter.memory-limit-bytes", 0, "Specifies the memory limit above which memory is considered to be under pressure.")
+	f.Uint64Var(&cfg.ResetBytes, prefix+"memory-limiter.reset-bytes", 0, "Specifies the memory limit below which full operation is resumed.")
+}
+
+// Interface for the memory limiter.
+type MemoryLimiter interface {
+	// Removes a request from the queue.
+	RemoveRequest(id int64)
+	// Adds a request to the queue.
+	AddRequest(id int64)
+	// If memory pressure is detected, waits will block until the memory pressure is removed or the context is done.
+	Wait(ctx context.Context, id int64)
+}
+
+type memLimiterCtxKey struct{}
+type requestIDCtxKey struct{}
+
+var (
+	mlCtxKey    = &memLimiterCtxKey{}
+	mlRidCtxKey = &requestIDCtxKey{}
+)
+
+// NewMemoryLimiter returns a new memorylimiter processor.
+func NewMemoryLimiter(cfg *MemoryLimiterConfig, logger log.Logger) (MemoryLimiter, error) {
+	if !cfg.Enabled {
+		return &NoOpMemoryLimiter{}, nil
+	}
+
+	if cfg.CheckInterval <= 0 {
+		return nil, errCheckIntervalOutOfRange
+	}
+	if cfg.MemoryLimitBytes == 0 {
+		return nil, errLimitOutOfRange
+	}
+	if cfg.MemoryLimitBytes == 0 {
+		return nil, errLimitOutOfRange
+	}
+
+	ml := &HeapMemoryLimiter{
+		memCheckInterval: cfg.CheckInterval,
+		memLimit:         cfg.MemoryLimitBytes,
+		resetLimit:       cfg.ResetBytes,
+		ticker:           time.NewTicker(cfg.CheckInterval),
+		readMemStatsFn:   runtime.ReadMemStats,
+		signalCh:         make(chan struct{}),
+		logger:           logger,
+		memPressure:      &atomic.Bool{},
+	}
+
+	ml.start()
+	return ml, nil
+}
+
+type HeapMemoryLimiter struct {
+	// The limit in bytes above which the system is considered to be under pressure.
+	memLimit uint64
+	// The limit above which the system is considered healthy.
```

Comment:
I believe this comment is not accurate. Do you mean "The limit below which"?

Reply:
@alvinlin123 - I addressed your comments. Could you take another look at the PR?